logger/logger.go

```diff
@@ -1,80 +1,173 @@
+	}
+	for _, opt := range opts {
+		opt(l)
+	}
+
+	if sentryDsn := os.Getenv("SENTRY_DSN"); sentryDsn != "" {
+		l.Info("SENTRY_DSN set, configuring SentryLogger", nil)
+		return NewSentryLogger(l, sentryDsn)
+	}
+
+	return l
+}
 
 // Debug writes a debug log.
-func (l *TrailsLogger) Debug(msg string, data map[string]interface{}) {
-	if data == nil {
-		l.l.Println(colorWhite + msg + colorClose)
+func (l *TrailsLogger) Debug(msg string, ctx *LogContext) {
+	if l.ll > LogLevelDebug {
 		return
 	}
-	l.l.Printf(colorWhite+msg+" %v"+colorClose+"\n", data)
+
+	l.log(color.WhiteString, LogLevelDebug, msg, ctx)
 }
 
-// Debug writes an error log.
-func (l *TrailsLogger) Error(msg string, data map[string]interface{}) {
-	if data == nil {
-		l.l.Println(colorRed + msg + colorClose)
+// Error writes an error log.
+func (l *TrailsLogger) Error(msg string, ctx *LogContext) {
+	if l.ll > LogLevelError {
 		return
 	}
-	l.l.Printf(colorRed+msg+" %v"+colorClose+"\n", data)
+
+	l.log(color.RedString, LogLevelError, msg, ctx)
 }
 
-// Debug writes a fatal log.
-func (l *TrailsLogger) Fatal(msg string, data map[string]interface{}) {
-	if data == nil {
-		l.l.Println(colorPink+msg+" %v"+colorClose, data)
+// Fatal writes a fatal log.
+func (l *TrailsLogger) Fatal(msg string, ctx *LogContext) {
+	if l.ll > LogLevelFatal {
 		return
 	}
-	l.l.Printf(colorPink+msg+" %v"+colorClose+"\n", data)
+
+	l.log(color.MagentaString, LogLevelFatal, msg, ctx)
 }
 
 // Info writes an info log.
-func (l *TrailsLogger) Info(msg string, data map[string]interface{}) {
-	if data == nil {
-		l.l.Println(colorBlue + msg + colorClose)
+func (l *TrailsLogger) Info(msg string, ctx *LogContext) {
+	if l.ll > LogLevelInfo {
 		return
 	}
-	l.l.Printf(colorBlue+msg+" %v"+colorClose+"\n", data)
+
+	l.log(color.BlueString, LogLevelInfo, msg, ctx)
 }
 
 // Warn writes a warning log.
-func (l *TrailsLogger) Warn(msg string, data map[string]interface{}) {
-	if data == nil {
-		l.l.Println(colorYellow + msg + colorClose)
+func (l *TrailsLogger) Warn(msg string, ctx *LogContext) {
+	if l.ll > LogLevelWarn {
 		return
 	}
-	l.l.Printf(colorYellow+msg+" %v"+colorClose+"\n", data)
+
+	l.log(color.YellowString, LogLevelWarn, msg, ctx)
+}
+
+// LogLevel returns the LogLevel set for the TrailsLogger.
+func (l *TrailsLogger) LogLevel() LogLevel { return l.ll }
+
+/*
+// WithContext includes the provided LogContext in the next log.
+func (l *TrailsLogger) WithContext(ctx LogContext) Logger {
+	logger := new(TrailsLogger)
+	*logger = *l
+	logger.ctx = ctx
+	return logger
+}
+*/
+
+// log executes printing the log message,
+// including any context if available.
+func (l *TrailsLogger) log(colorizer func(string, ...interface{}) string, level LogLevel, msg string, ctx *LogContext) {
+	// TODO(dlk): have skip be configurable or implement some logic
+	// like https://github.com/sirupsen/logrus/blob/b50299cfaaa1bca85be76c8984070e846c7abfd2/entry.go#L178-L213
+	_, file, line, _ := runtime.Caller(l.skip)
```

Comment:
@profsmallpine I think your question about stack trace is caller file & line. Because a logger is instantiated app-wide, there's only one skip count we can use. But, given there are different depths of the call site in the stack, the sample code referenced here shows a dynamic approach that would be a good reference; essentially, get to the caller of the first package that is outside `trails`.